File: C++ Projects/DSA/Perfect_Square.cpp
#include <bits/stdc++.h>

using namespace std;

int main()
{
    int num, half;
    cin >> num;

    half = num / 2;

    int start = 1, mid, end = half;

    while (start - end != -1 && start - end != 1)
    {
        mid = (start + end) / 2;

        if (start * start == num)
        {
            cout << start;
            break;
        }

        if (end * end == num)
        {
            cout << end;
            break;
        }

        if (mid * mid > num)
        {
            end = mid;
        }

        else if (mid * mid < num)
        {
            start = mid;
        }

        else if (mid * mid == num)
        {
            cout << mid;
            break;
        }
    }
}

File: JAVA Projects/LeetCode/Perfect_Square.java
package LeetCode;

public class Perfect_Square {
    public static boolean isPerfectSquare(int num) {
        int start,mid,end;
        start=1;end=num/2+1;
        while(start<=end){
            mid=start+(end-start)/2;
            if(mid==num/mid && mid*mid ==num){
                return true;
            }else if(mid>num/mid){
                end = mid-1;
            }else if(mid<num/mid){
                start = mid+1;
            }else break;
        }
        return false;
    }
    public static void main(String[] args) {
        int num = 5;
        isPerfectSquare(num);
    }
}
